tests/unit/fem/ptest_tdofs.cpp

```diff
@@ -0,0 +1,1012 @@
+#endif //PTEST_TDOF_USE_DEBUG_PRINT
+
+// Define some parameters for the tests
+const int NUM_SER_REF_LEVELS = 1;
+const int NUM_PAR_REF_LEVELS = 0;
+const bool VERBOSE_OUTPUT = false;
+
+/// Helper function to refine the mesh uniformly
+inline void refineMesh(mfem::Mesh* mesh, int num_levels)
+{
+   for (int l = 0; l < num_levels; ++l)
+   {
+      mesh->UniformRefinement();
+   }
+}
+
+/// Helper function to collect some stats about the mesh
+inline std::string getMeshStats(mfem::Mesh* mesh)
+{
+   std::stringstream sstr;
+
+   sstr << "Mesh has: "
+        << "\n\t"<< mesh->GetNV() << " vertices."
+        << "\n\t"<< mesh->GetNEdges() << " edges."
+        << "\n\t"<< mesh->GetNFaces() << " faces."
+        << "\n\t"<< mesh->GetNE() << " elements."
+        << "\n\t"<< mesh->GetNBE() << " boundary elements.";
+
+   return sstr.str();
+}
+
+/// Helper to remove encoded sign from a DOF
+/// Adapted from protected static function in mfem::FiniteElementSpace
+inline int decodeDof(int dof)
+{
+   return (dof >= 0) ? dof: (-1 - dof);
+}
+
+/// Helper class for testing true dofs on parmesh elements
+struct TDOFTester
+{
+   TDOFTester(mfem::ParMesh* mesh, bool verbose)
+      : m_mesh(mesh)
+      , m_verbose(verbose)
+   {
+      rank = m_mesh->GetMyRank();
+
+      MPRINT(rank, "Details for parallel mesh:\n" << getMeshStats(m_mesh));
+   }
+
+   /// Given an array of TDofs, apply some checks
+   /// typeStr parameter is only used for logging
+   void checkDOFs(const mfem::Array<int>& arr, const std::string& typeStr)
+   {
+      MPRINT_MASTER(rank,
+                    "List of "<< typeStr << " tdofs in pmesh\n"
+                    << "  -----------------------------");
+
+      //const bool isNC = m_mesh->Nonconforming();
+      //const int CHECK_VAL = isNC ? -1 : 0;
+      const int CHECK_VAL = 0;
+
+      std::stringstream sstr;
+      for (int i=0; i< arr.Size(); ++i)
+      {
+         // Check that the TDOF is non-negative
+         CHECK( arr[i] >= CHECK_VAL);
+
+         sstr << "\n\t" << typeStr << " "
+              << i << " -> " << arr[i];
+      }
+      MPRINT(rank, sstr.str());
+
+      MPRINT_MASTER(rank,"-----------------------------");
+   }
+
+   /// Generate and test tdofs for the mesh vertices
+   void testVertexTDofs()
+   {
+      const int dim = m_mesh->Dimension();
+      const int order = 1;
+
+      const int NV = m_mesh->GetNV();
+
+      MPRINT_MASTER(rank,"Finding vertex tdofs");
+
+      auto* fec = new mfem::H1_FECollection(order, dim);
+      mfem::ParFiniteElementSpace fes(m_mesh, fec);
+      mfem::Array<int> vert_tdof(NV);
+      vert_tdof = UNITIALIZED_VAL;
+
+      mfem::Array<int> idofs;
+
+      std::stringstream sstr;
+      for (int v=0; v< NV; ++v)
+      {
+         fes.GetVertexDofs(v, idofs);
+         CHECK( idofs.Size() == 1);
+
+         auto ldof = idofs[0];
+         CHECK( ldof >= 0);
+
+         // WARNING -- The following does not account for NCMesh!
```

Comment:
I couldn't figure out how to make this work for non-conforming meshes.

``ParFiniteElementSpace::GetGlobalTDofNumber(ldof)`` has an ``MFEM_VERIFY`` that the tdof is not ``-1``.
Following @v-dobrev's suggestion in #1102, I tried using ``ParFiniteElementSpace::GetLocalTDofNumber(ldof)`` first to see if the element was local, but that didn't work, since it returns ``-1`` for any element that isn't owned.

Basically, I think there are two ideas here that we need to resolve (and apologies if I'm not using mfem's preferred nomenclature):
* One is the distinction between the owned and non-owned elements of the mesh that the ``ParFiniteElementSpace`` takes care of for us -- essentially, the distinction between the global and local tdofs.
* The other, for nc-meshes, is the distinction between the master and slave nodes (i.e. the hanging nodes). Is there a function to help figure out what the true dofs are for these?

Relevant snippets of code:
* https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/pfespace.cpp#L765-L784
* https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/pfespace.cpp#L786-L807
* Would it help if ``GetGlobalTDofNumber`` called ``Dof_TrueDof_Matrix()``?

Reply:
If the goal is to get global numbering for vertices, including slave vertices, I think the simplest approach would be to coerce `ParFiniteElementSpace` to call `BuildParallelConformingInterpolation` with the option `partial = true` in the constructor. This is a feature that was added for hybridization, though it is not user accessible, so the Space class would have to be modified. The effect is that only a partially-conforming space is constructed, where the parallel decomposition is handled as usual between conforming entities, but master-slave interfaces stay disconnected. As a result, slaves should obtain true DOF numbers.

Reply:
Thanks @jakubcerveny.

I don't see a constructor with the ``partial`` option, but it looks like ``BuildParallelConformingInterpolation()`` gets called by: ``ParFiniteElementSpace::GetPartialConformingInterpolation()``
See: https://github.com/mfem/mfem/blob/c55c80d17b82d80de04b849dd526e17044f8c99a/fem/pfespace.cpp#L667-L676

Is this safe to call on a ``ParFESpace`` over an ``NCMesh``?

> but master-slave interfaces stay disconnected

Could you please clarify this?  If there is a slave vertex on two different ranks will they get the same global index after calling ``BuildParallelConformingInterpolation()`` ?

Also, will calling ``BuildParallelConformingInterpolation()`` make it safe to call  ``GetGlobalTDofNumber()`` on all the vertices (and other faces) for an ``NCMesh``?
I.e. that it will no longer trigger the contained ``MFEM_VERIFY``.

Reply:
On line 174 in `pfespace.cpp`, there would have to be a `true` at the end, that's the modification of the space that I had in mind. The method `GetPartialConformingInterpolation` is the special hybridization entrypoint and is not relevant here. 

After the modification, slave DOFs will become true DOFs and will not interpolate the master edge/face (i.e., AMR will be broken). But, `GetGlobalTDofNumber` should return a global index for a slave DOF that is owned by the current rank. 

This will probably not work for slaves that are owned by a neighboring rank, these DOFs will still have -1 in the `ldof_ltdof` vector on the non-owning ranks, because `BuildParallelConformingInterpolation` messes with it. Can you try commenting out lines 173-174, and 788-794 in `prefspace.cpp`?

Reply:
OK, I tried that and it doesn't work because `tdof_nb_offsets` is not set. Making it work as you need would require more changes...

Reply:
I'm not sure what the application is and whether an "ultimate" solution is needed or a hacky solution would suffice. It seems the following hack should work. Before constructing the `ParFiniteElementSpace`, try swapping out `pmesh->pncmesh` with `NULL`:
```
ParNCMesh *pncmesh = NULL;
std::swap(pmesh->pncmesh, pncmesh);
```
Then construct the parallel space, it will function as if the mesh was conforming. Slaves will become true DOFs and `GetGlobalTDofNumber` will work. After reading out the global indices, destroy the space and swap the `pncmesh` pointer back.

Reply:
Thanks @jakubcerveny, I can try that.  
Are there any side effects of this? E.g. on internal ``ParMesh`` state variables.

It would be best for me if we had more of a robust solution that is not likely to break since this is a long term need. This is part of the reason I've added the comprehensive unit test to mfem.

We are developing a contact mechanics library that computes forces and other fields on the boundary surface (vertices and elements). The global IDs on the vertices and faces are needed for a thirdparty domain partitioner on the boundary elements of my mesh, which allows the serial contact library to interface with a parallel mesh (and to distribute the work).

Reply:
For long term needs (or if the trick fails), I can develop a special function in NCMesh that will assign the global vertex IDs exactly as you need. We can prioritize with @tzanio.